### PR TITLE
support ruby and better raw html display

### PR DESCRIPTION
**Pull request by Jocs**
| Q                | A
| ---------------- | ---
| Bug fix?         | yes
| New feature?     | yes
| BC breaks?       | no
| Deprecations?    | no
| New tests added? | not needed
| Fixed tickets    | #257
| License          | MIT

### Description

**auto-complete**

1. write `span#id.className`
2. press <kbd>tab</kbd>.

write markdown below:

```md
<span style="color: red">red text</span>
<kbd>Ctrl</kbd>+<kbd>C</kbd>
<ruby> 漢 <rp>(</rp><rt>Kan</rt><rp>)</rp> 字 <rp>(</rp><rt>ji</rt><rp>)</rp></ruby>
```

will look like below:

<span style="color: red">red text</span>
<kbd>Ctrl</kbd>+<kbd>C</kbd>
<ruby> 漢 <rp>(</rp><rt>Kan</rt><rp>)</rp> 字 <rp>(</rp><rt>ji</rt><rp>)</rp></ruby>

屏幕快照 2019-04-03 下午1 17 56

### CODE REVIEW

File: src/muya/lib/parser/parse.js [hunk 1]
```diff
@@ -16,21 +17,21 @@ const validWidthAndHeight = value => {
...
+const getAttributes = html => {
+  const parser = new DOMParser()
+  const doc = parser.parseFromString(html, 'text/html')
+  const target = doc.querySelector('body').firstElementChild
+  if (!target) return null
+  const attrs = {}
+  for (const attr of target.getAttributeNames()) {
+    if (/width|height/.test(attr)) {
+      attrs[attr] = validWidthAndHeight(target.getAttribute(attr))
+    } else {
```
Comment (fxha) [inline src/muya/lib/parser/parse.js]: `} else if (!global.MARKTEXT_SAFE_MODE) {`

If not `src`, `alt`, `width` or `height`.

**Comment by fxha** [comment]
@Jocs Do you want this feature in v0.14.0?

**Comment by Jocs** [comment]
>span#id.className + <tab> does nothing

fixed,

>Your example is displayed in one line when exported/printed but Mark Text displays it as three lines

I think this is right, Because only `hard line break` will export a `<br />`. https://github.github.com/gfm/#hard-line-breaks

**Comment by Jocs** [comment]
>} else if (!global.MARKTEXT_SAFE_MODE) {
>
>If not src, alt, width or height.

Sorry I didn't understand the meaning.

**Comment by Jocs** [comment]
>I would add a settings option to enable styled HTML and default disable styled elements due security reasons and all markdown parsers like GitHub skip styled elements. Also I found two issues:

I think we can add this in the next version.

**Comment by Jocs** [comment]
> Do you want this feature in v0.14.0?

Yes.

**Comment by fxha** [comment]
> Sorry I didn't understand the meaning.

You should only enable custom (not safe attributes) when `global.MARKTEXT_SAFE_MODE` is not defined because:

> due security reasons and all markdown parsers like GitHub skip styled elements

**Comment by fxha** [comment]
> I think this is right, Because only hard line break will export a `<br />`. https://github.github.com/gfm/#hard-line-breaks

Then it's a muya/Mark Text bug?!

**Comment by Jocs** [comment]
>You should only enable custom (not safe) attributes when `global.MARKTEXT_SAFE_MODE` is not defined because:

Yes, I think we need to filter custom attributes. NO matter whether `global.MARKTEXT_SAFE_MODE` is defined. I'll modify this.

**Comment by Jocs** [comment]
>Then it's a muya/Mark Text bug?!

I think it's not a bug, because we display both `soft line break` and `hard line break` in different lines, and use a `bended arrow` to distinguish them.

屏幕快照 2019-04-04 上午12 51 05

the line break after `red text` is a soft line break, and the line break after <kbd>C</kbd> is a hard line break.

**Comment by Jocs** [comment]
@fxha I have a question, does `style` attribute has security risk? why GitHub SKIP it?

**Comment by fxha** [comment]
> I have a question, does style attribute has security risk? why GitHub SKIP it?

You could manipulate the page e.g. with `absolute`/`fixed` elements etc.

**Pull request merged**